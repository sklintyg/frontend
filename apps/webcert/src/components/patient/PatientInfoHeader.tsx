import type React from 'react'
import { useDispatch } from 'react-redux'
import { useHistory } from 'react-router-dom'
import styled from 'styled-components'
import { swapImage, userImage } from '../../images'
import { clearPatient } from '../../store/patient/patientActions'
import { BoxShadowContainer } from '../../styles'
import type { Patient } from '../../types'
import { CustomButton } from '../Inputs/CustomButton'
import PatientStatuses from '../notification/PatientStatuses'

interface Props {
  patient: Patient
}

const ButtonWrapper = styled.div`
  margin-top: auto;
`

const PatientInfoHeader: React.FC<Props> = ({ patient }) => {
  const dispatch = useDispatch()
  const history = useHistory()

  const onSwitchPatient = () => {
    dispatch(clearPatient())
    history.push('/search')
  }

  return (
    <BoxShadowContainer>
      <div className="ic-container iu-flex iu-py-400">
        <img src={userImage} alt="patient" className="iu-mr-300" />
        <div>
          <h3 className="iu-mb-200">Patientuppgifter</h3>
          <div className="iu-flex">
            <h2>
              {patient.fullName} - {patient.personId.id}
            </h2>
            <ButtonWrapper>
              <CustomButton
                text="Byt patient"
                buttonStyle="primary"
                onClick={onSwitchPatient}
                className="iu-ml-500"
                tooltip="Byt patient att skriva och söka intyg för."
                startIcon={<img src={swapImage} alt="Byt patient" />}
              />
            </ButtonWrapper>
          </div>
          <PatientStatuses patient={patient} />
        </div>
      </div>
    </BoxShadowContainer>
  )
}
export default PatientInfoHeader
