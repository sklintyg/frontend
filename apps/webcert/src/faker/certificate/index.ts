export * from './fakeCertificate'
export * from './fakeCertificateConfig'
export * from './fakeCertificateData'
export * from './fakeCertificateDataValidation'
export * from './fakeCertificateMetaData'
export * from './fakeCertificateRelation'
export * from './fakeCertificateValue'
