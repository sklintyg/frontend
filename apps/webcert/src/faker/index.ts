export * from './certificate'
export * from './fakeCity'
export * from './fakeList'
export * from './fakePatient'
export * from './fakeResourceLink'
export * from './fakeSrs'
export * from './fakeStaff'
export * from './icf'
export * from './list'
export * from './question'
export * from './user'
