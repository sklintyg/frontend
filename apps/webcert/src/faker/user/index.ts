export * from './fakeCareProvider'
export * from './fakeUnit'
export * from './fakeUnitStatistic'
export * from './fakeUser'
export * from './fakeUserStatistics'
