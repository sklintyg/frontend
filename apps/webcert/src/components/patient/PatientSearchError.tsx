import type React from 'react'
import { useSelector } from 'react-redux'
import styled from 'styled-components'
import { ErrorCode } from '../../store/error/errorReducer'
import { getPatientError } from '../../store/patient/patientSelectors'
import WCDynamicLink from '../../utils/WCDynamicLink'
import ErrorCopyText from '../error/ErrorCopyText'
import InfoBox from '../utils/InfoBox'

const Wrapper = styled.div`
  max-width: 720px;
`

const PatientSearchError = () => {
  const error = useSelector(getPatientError)

  if (!error) {
    return null
  }

  const getContent = () => {
    if (error.errorCode === ErrorCode.GETTING_PATIENT_ERROR) {
      return (
        <p>
          På grund av tekniskt fel gick det inte att hämta personuppgifter, försök igen om en liten stund. Om problemet kvarstår, kontakta i
          första hand din lokala IT-avdelning och i andra hand <WCDynamicLink linkKey={'ineraKundserviceAnmalFel'} />.
        </p>
      )
    } else if (error.errorCode === ErrorCode.PATIENT_NOT_FOUND) {
      return (
        <p>
          Personnumret du har angivit finns inte i folkbokföringsregistret. Kontrollera om du har skrivit rätt. Observera att det inte går
          att ange reservnummer. Webcert hanterar enbart person- och samordningsnummer.
        </p>
      )
    } else if (error.errorCode === ErrorCode.INVALID_PATIENT_ID) {
      return <p>Ange ett giltigt person- eller samordningsnummer.</p>
    } else if (error.errorCode === ErrorCode.PU_PROBLEM) {
      return (
        <>
          <p>
            Personuppgiftstjänsten svarar inte. Åtgärden kan inte genomföras eftersom den kräver att personuppgifter kan hämtas från
            personuppgiftsregistret. Prova igen om en stund.
          </p>
          <p>
            Om problemet kvarstår, kontakta i första hand din lokala IT-avdelning och i andra hand{' '}
            <WCDynamicLink linkKey={'ineraKundserviceAnmalFel'} />.
          </p>
        </>
      )
    } else if (error.errorCode === ErrorCode.PATIENT_NO_NAME) {
      return <p>Förnamn eller efternamn för det personnummer du har angett kunde tyvärr inte hämtas från folkbokföringsregistret.</p>
    }
  }

  return (
    <Wrapper>
      <InfoBox type="error" additionalStyles="iu-mt-300" activateIconWrap>
        {getContent()}
      </InfoBox>
      {error.errorId && (error.errorCode === ErrorCode.GETTING_PATIENT_ERROR || error.errorCode === ErrorCode.PU_PROBLEM) && (
        <ErrorCopyText errorId={error.errorId} />
      )}
    </Wrapper>
  )
}

export default PatientSearchError
