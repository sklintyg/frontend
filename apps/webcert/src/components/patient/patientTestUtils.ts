import { fakePatient } from '../../faker'
import { Patient } from '../../types'

/** @deprecated use fakePatient() */
export const createPatient = (patientId: string): Patient => {
  return fakePatient({
    firstName: 'firstName',
    lastName: 'lastName',
    fullName: 'firstName middleName lastName',
    street: 'Street 1',
    zipCode: '12345',
    city: 'City',
    personId: {
      type: '',
      id: patientId,
    },
  })
}
