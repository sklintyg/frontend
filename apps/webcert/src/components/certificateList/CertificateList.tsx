import { useEffect, useState } from 'react'
import { useDispatch } from 'react-redux'
import { useHistory } from 'react-router-dom'
import ReactTooltip from 'react-tooltip'
import styled from 'styled-components'
import { fileImage } from '../../images'
import { updateCreatedCertificateId } from '../../store/certificate/certificateActions'
import { getCertificateId } from '../../store/certificate/certificateSelectors'
import { getCertificateTypes } from '../../store/patient/patientActions'
import { getActivePatient, selectCertificateTypes } from '../../store/patient/patientSelectors'
import { useAppSelector } from '../../store/store'
import { setUserPreference } from '../../store/user/userActions'
import { getUserPreference } from '../../store/user/userSelectors'
import { selectIsLoadingInitialState } from '../../store/utils/utilsSelectors'
import Spinner from '../utils/Spinner'
import { CertificateListRow } from './CertificateListRow'

const sortByFavorite = (a: boolean, b: boolean): number => {
  if (a > b) {
    return -1
  } else if (a < b) {
    return 1
  } else {
    return 0
  }
}

const CertificateBox = styled.div`
  max-height: 30.2em;
  overflow-y: auto;
`

const FlexWrapper = styled.div`
  flex: 1;
`

export function CertificateList() {
  const certificateId = useAppSelector(getCertificateId())
  const userPreferences = useAppSelector(getUserPreference('wc.favoritIntyg'))
  const certificateTypes = useAppSelector(selectCertificateTypes)
  const patient = useAppSelector(getActivePatient)
  const isLoadingInitialState = useAppSelector(selectIsLoadingInitialState)

  const [favorites, setFavorites] = useState<string[]>([])

  const dispatch = useDispatch()
  const history = useHistory()

  useEffect(() => {
    ReactTooltip.hide()
  }, [favorites])

  const handlePreferenceClick = (id: string) => {
    let updatedFavorites = []

    if (favorites.includes(id)) {
      updatedFavorites = favorites.filter((f) => f !== id)
    } else {
      updatedFavorites = [...favorites, id]
    }

    setFavorites(updatedFavorites)
    dispatch(setUserPreference({ key: 'wc.favoritIntyg', value: JSON.stringify(updatedFavorites) }))
  }

  useEffect(() => {
    if (patient) {
      dispatch(getCertificateTypes(patient.personId.id))
    }
  }, [dispatch, patient])

  useEffect(() => {
    if (userPreferences) {
      setFavorites([...JSON.parse(userPreferences)])
    }
  }, [userPreferences])

  useEffect(() => {
    if (certificateId) {
      dispatch(updateCreatedCertificateId(''))
      history.push(`/certificate/${certificateId}`)
    }
  }, [certificateId, dispatch, history])

  return (
    <div className="iu-flex">
      <div className="iu-mr-gutter">
        <img src={fileImage} alt="Ikon fÃ¶r skapa intyg" className="iu-height-600" />
      </div>
      <FlexWrapper>
        <h3 className="iu-mb-05rem">Skapa intyg</h3>
        {isLoadingInitialState ? (
          <Spinner />
        ) : (
          <CertificateBox className="iu-border-secondary-light iu-shadow-sm iu-flex iu-flex-column">
            {[...certificateTypes]
              .sort(({ id: a }, { id: b }) => sortByFavorite(favorites.includes(a), favorites.includes(b)))
              .map(({ label, detailedDescription, id, issuerTypeId, links, message, confirmationModal }) => {
                return (
                  <CertificateListRow
                    certificateName={label}
                    certificateInfo={detailedDescription}
                    id={id}
                    issuerTypeId={issuerTypeId}
                    favorite={favorites.includes(id)}
                    preferenceClick={handlePreferenceClick}
                    message={message}
                    key={id}
                    patient={patient}
                    links={links ?? []}
                    confirmationModal={confirmationModal}
                  />
                )
              })}
          </CertificateBox>
        )}
      </FlexWrapper>
    </div>
  )
}
