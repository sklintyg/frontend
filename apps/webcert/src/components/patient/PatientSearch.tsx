import type React from 'react'
import { useCallback, useEffect, useState } from 'react'
import { useHistory } from 'react-router-dom'
import styled from 'styled-components'
import { clearPatientError, getPatient } from '../../store/patient/patientActions'
import { getActivePatient } from '../../store/patient/patientSelectors'
import { useAppDispatch, useAppSelector } from '../../store/store'
import { isPersonIdValid, useKeyPress } from '../../utils'
import { CustomButton } from '../Inputs/CustomButton'
import PersonIdInput from '../Inputs/PersonIdInput'
import PatientSearchError from './PatientSearchError'

const FormWrapper = styled.div`
  display: flex;
  align-items: center;
`

const PatientSearch: React.FC = () => {
  const [patientId, setPatientId] = useState('')
  const patient = useAppSelector(getActivePatient)
  const dispatch = useAppDispatch()
  const enterPress = useKeyPress('Enter')
  const history = useHistory()

  const onSubmit = useCallback(() => {
    dispatch(getPatient(patientId))
  }, [dispatch, patientId])

  useEffect(() => {
    if (enterPress && isPersonIdValid(patientId)) {
      onSubmit()
    }
  }, [enterPress, onSubmit, patientId])

  useEffect(() => {
    if (patient) {
      history.push(`/create/${encodeURIComponent(btoa(patientId))}`)
    }
  }, [patient, history, patientId])

  const onChange = (formattedPatientId: string) => {
    dispatch(clearPatientError())
    setPatientId(formattedPatientId)
    if (formattedPatientId.length === 0) {
      onFocus()
    }
  }

  const onFocus = () => {
    dispatch(clearPatientError())
  }

  return (
    <>
      <h2>Patientens personnummer eller samordningsnummer</h2>
      <FormWrapper className="iu-mt-300">
        <PersonIdInput onFormattedChange={onChange} value={patientId} onFocus={onFocus} />
        <CustomButton
          text="Fortsätt"
          disabled={!isPersonIdValid(patientId)}
          buttonStyle="primary"
          onClick={onSubmit}
          tooltip="Gå vidare för att skapa eller söka efter intyg för patienten."
          inline={true}
        />
      </FormWrapper>
      <PatientSearchError />
    </>
  )
}
export default PatientSearch
