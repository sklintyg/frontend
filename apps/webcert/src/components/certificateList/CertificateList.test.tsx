import type { EnhancedStore } from '@reduxjs/toolkit'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { createBrowserHistory } from 'history'
import { Provider } from 'react-redux'
import { Router } from 'react-router-dom'
import { vi } from 'vitest'
import { fakePatient, fakeResourceLink } from '../../faker'
import { updateCreatedCertificateId } from '../../store/certificate/certificateActions'
import { configureApplicationStore } from '../../store/configureApplicationStore'
import { setPatient, updateCertificateTypes } from '../../store/patient/patientActions'
import { patientMiddleware } from '../../store/patient/patientMiddleware'
import dispatchHelperMiddleware, { clearDispatchedActions } from '../../store/test/dispatchHelperMiddleware'
import { updateUser, updateUserPreference } from '../../store/user/userActions'
import { userMiddleware } from '../../store/user/userMiddleware'
import type { CertificateType, ResourceLink, User } from '../../types';
import { ResourceLinkType } from '../../types'
import CertificateList from './CertificateList'

const createType = ({
  description = '',
  detailedDescription = '',
  id = '',
  issuerTypeId = '',
  label = '',
  links = [],
  message = '',
}: {
  description?: string
  detailedDescription?: string
  id?: string
  issuerTypeId?: string
  label?: string
  links?: ResourceLink[]
  message?: string
}): CertificateType => ({
  description,
  detailedDescription,
  id,
  issuerTypeId,
  label,
  links,
  message,
})

let testStore: EnhancedStore
let types: CertificateType[]
const testHistory = createBrowserHistory()
testHistory.push = vi.fn()

const renderComponent = () =>
  render(
    <Provider store={testStore}>
      <Router history={testHistory}>
        <CertificateList />
      </Router>
    </Provider>
  )

describe('CertificateList', () => {
  beforeEach(() => {
    testStore = configureApplicationStore([dispatchHelperMiddleware, userMiddleware, patientMiddleware])

    types = [
      createType({
        id: 'typ1',
        label: 'Typ 1',
        message: 'Meddelande',
        links: [
          fakeResourceLink({ type: ResourceLinkType.CREATE_CERTIFICATE, name: 'Skapa intyg' }),
          fakeResourceLink({ type: ResourceLinkType.MISSING_RELATED_CERTIFICATE_CONFIRMATION }),
        ],
      }),
      createType({
        id: 'typ2',
        label: 'Typ 2',
        links: [
          fakeResourceLink({ type: ResourceLinkType.CREATE_CERTIFICATE, name: 'Skapa intyg' }),
          fakeResourceLink({ type: ResourceLinkType.CREATE_DODSBEVIS_CONFIRMATION, name: 'Dödsbevis saknas' }),
        ],
      }),
      createType({ id: 'typ3', label: 'Typ 3' }),
      createType({ id: 'typ4', label: 'Typ 4' }),
      createType({ id: 'typ5', label: 'Typ 5' }),
      createType({
        id: 'typ6',
        label: 'Typ 6',
        links: [
          fakeResourceLink({ type: ResourceLinkType.CREATE_CERTIFICATE, name: 'Skapa intyg' }),
          fakeResourceLink({ type: ResourceLinkType.CREATE_LUAENA_CONFIRMATION, name: 'Luaena saknas' }),
        ],
      }),
    ]

    testStore.dispatch(updateCertificateTypes(types))
  })

  afterEach(() => {
    clearDispatchedActions()
  })

  const dispatchFavorites = (favorites: string[]) => {
    testStore.dispatch(updateUser({ preferences: null } as User))
    testStore.dispatch(updateUserPreference({ key: 'wc.favoritIntyg', value: JSON.stringify(favorites) }))
  }

  it('should render list of certificate types', () => {
    renderComponent()

    const labels = screen.getAllByText('Typ', { exact: false }).map((el) => el.textContent?.trim())
    expect(labels).toEqual(types.map((t) => t.label))
  })

  it('should order favorite certificate types first', () => {
    dispatchFavorites(['typ2', 'typ4'])
    renderComponent()

    const labels = screen.getAllByText('Typ', { exact: false }).map((el) => el.textContent?.trim())
    expect(labels).toEqual(['Typ 2', 'Typ 4', 'Typ 1', 'Typ 3', 'Typ 5', 'Typ 6'])
  })

  it('should add favorites', async () => {
    renderComponent()

    const buttons = screen.getAllByLabelText('Markera intyget som favorit och fäst högst upp i listan.')
    await userEvent.click(buttons[1])

    expect(buttons[1]).toMatchSnapshot()
  })

  it('should remove favorites', async () => {
    renderComponent()

    const buttons = screen.getAllByLabelText('Markera intyget som favorit och fäst högst upp i listan.')
    await userEvent.click(buttons[1])
    await userEvent.click(buttons[1])

    expect(buttons[1]).toMatchSnapshot()
  })

  it('should show modal when clicked', async () => {
    renderComponent()

    const aboutLink = screen.getAllByText('Om intyget')[0] as HTMLElement
    await userEvent.click(aboutLink)

    expect(screen.getByRole('dialog')).toBeInTheDocument()
  })

  it('should redirect user when certificate id is set', () => {
    testStore.dispatch(updateCreatedCertificateId('certificateId'))

    renderComponent()

    expect(testHistory.push).toHaveBeenCalledWith('/certificate/certificateId')
  })

  it('should clear certificate id after certificate id is set', () => {
    testStore.dispatch(updateCreatedCertificateId('certificateId'))

    renderComponent()

    expect(testStore.getState().ui.uiCertificate.createdCertificateId).toEqual('')
  })

  it('should show message when it exists', () => {
    renderComponent()

    expect(screen.getByText('Meddelande')).toBeInTheDocument()
  })

  it('should show confirm modal when CREATE_DODSBEVIS_CONFIRMATION resource link exists', async () => {
    testStore.dispatch(setPatient(fakePatient()))
    renderComponent()

    const button = screen.getAllByRole('button', {
      name: /Skapa intyg/,
    }) as HTMLElement[]
    await userEvent.click(button[1])

    expect(screen.getByText('Du är på väg att utfärda ett dödsbevis för', { exact: false })).toBeInTheDocument()
  })

  it('should show confirm modal when CREATE_LUAENA_CONFIRMATION resource link exists', async () => {
    testStore.dispatch(setPatient(fakePatient()))
    renderComponent()

    const button = screen.getAllByRole('button', {
      name: /Skapa intyg/,
    }) as HTMLElement[]
    await userEvent.click(button[2])

    expect(screen.getByText('Du är på väg att utfärda Läkarutlåtande för', { exact: false })).toBeInTheDocument()
  })

  it('should show modal when MISSING_RELATED_CERTIFICATE_CONFIRMATION resource link exists', async () => {
    testStore.dispatch(setPatient(fakePatient()))
    renderComponent()

    const button = screen.getAllByRole('button', {
      name: /Skapa intyg/,
    })
    await userEvent.click(button[0])

    expect(screen.getByRole('dialog')).toBeInTheDocument()
  })
})
