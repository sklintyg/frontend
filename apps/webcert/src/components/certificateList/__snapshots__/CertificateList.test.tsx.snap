// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CertificateList > should add favorites 1`] = `
<button
  aria-label="Ta bort som favoritmarkerat intyg."
  class="sc-jMakVo eFBDMf iu-mr-1rem"
  data-tip="Ta bort som favoritmarkerat intyg."
>
  <svg
    aria-hidden="true"
    class="iu-color-information"
    fill="currentColor"
    focusable="false"
    style="display: inline-block; width: 1em; height: 1em; vertical-align: -0.125em;"
    viewBox="0 0 40 41"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M22.2981 2.59977C21.4479 0.464139 18.5521 0.46414 17.7019 2.59977L14.0004 11.8973C13.642 12.7977 12.8316 13.4128 11.9012 13.4907L2.29359 14.2954C0.0867348 14.4803 -0.808112 17.3576 0.873276 18.8623L8.19327 25.4132C8.9021 26.0476 9.21166 27.0429 8.9951 27.9914L6.75872 37.7863C6.24503 40.0362 8.58777 41.8144 10.4771 40.6088L18.7026 35.3599C19.4991 34.8516 20.5009 34.8516 21.2974 35.3599L29.5229 40.6088C31.4122 41.8144 33.755 40.0362 33.2413 37.7863L31.0049 27.9914C30.7883 27.0429 31.0979 26.0476 31.8067 25.4132L39.1267 18.8623C40.8081 17.3576 39.9133 14.4803 37.7064 14.2954L28.0988 13.4907C27.1684 13.4128 26.358 12.7977 25.9996 11.8973L22.2981 2.59977Z"
      fill="currentColor"
    />
  </svg>
</button>
`;

exports[`CertificateList > should remove favorites 1`] = `
<button
  aria-label="Markera intyget som favorit och fäst högst upp i listan."
  class="sc-jMakVo eFBDMf iu-mr-1rem"
  data-tip="Markera intyget som favorit och fäst högst upp i listan."
>
  <svg
    aria-hidden="true"
    class="iu-color-muted"
    fill="currentColor"
    focusable="false"
    style="display: inline-block; width: 1em; height: 1em; vertical-align: -0.125em;"
    viewBox="0 0 40 41"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      clip-rule="evenodd"
      d="M24.1414 12.6371L20.4399 3.33952C20.3764 3.17978 20.2971 3.10596 20.2389 3.06751C20.1726 3.02373 20.0899 2.99805 20 2.99805C19.9101 2.99805 19.8274 3.02373 19.7611 3.06751C19.7029 3.10596 19.6236 3.17978 19.56 3.33952L15.8586 12.6371C15.2331 14.2081 13.79 15.3396 12.0681 15.4838L2.46051 16.2885C2.34471 16.2982 2.27251 16.3361 2.21489 16.3861C2.14743 16.4446 2.07718 16.5426 2.03511 16.6779C1.94887 16.9552 2.02001 17.2046 2.20703 17.372L9.52703 23.9229C10.7932 25.056 11.3188 26.799 10.9449 28.4366L8.70854 38.2315C8.63743 38.543 8.75432 38.77 8.92599 38.9003C9.01401 38.9671 9.0989 38.9931 9.16195 38.9973C9.21238 39.0008 9.28875 38.9946 9.40128 38.9228L17.6268 33.6739C19.0794 32.7469 20.9206 32.7469 22.3732 33.6739L30.5987 38.9228C30.7113 38.9946 30.7876 39.0008 30.8381 38.9973C30.9011 38.9931 30.986 38.9671 31.074 38.9003C31.2457 38.77 31.3626 38.543 31.2915 38.2315L29.0551 28.4366C28.6812 26.799 29.2069 25.056 30.473 23.9229L37.793 17.372C37.98 17.2046 38.0511 16.9552 37.9649 16.6779C37.9228 16.5426 37.8526 16.4446 37.7851 16.3861C37.7275 16.3361 37.6553 16.2982 37.5395 16.2885L27.9319 15.4838C26.21 15.3396 24.7669 14.2081 24.1414 12.6371ZM22.2981 2.59977C21.4479 0.464139 18.5521 0.46414 17.7019 2.59977L14.0004 11.8973C13.642 12.7977 12.8316 13.4128 11.9012 13.4907L2.29359 14.2954C0.0867348 14.4803 -0.808112 17.3576 0.873276 18.8623L8.19327 25.4132C8.9021 26.0476 9.21166 27.0429 8.9951 27.9914L6.75872 37.7863C6.24503 40.0362 8.58777 41.8144 10.4771 40.6088L18.7026 35.3599C19.4991 34.8516 20.5009 34.8516 21.2974 35.3599L29.5229 40.6088C31.4122 41.8144 33.755 40.0362 33.2413 37.7863L31.0049 27.9914C30.7883 27.0429 31.0979 26.0476 31.8067 25.4132L39.1267 18.8623C40.8081 17.3576 39.9133 14.4803 37.7064 14.2954L28.0988 13.4907C27.1684 13.4128 26.358 12.7977 25.9996 11.8973L22.2981 2.59977Z"
      fill="currentColor"
      fill-rule="evenodd"
    />
  </svg>
</button>
`;
